agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,165 @@
+ */
+NR_PHP_WRAPPER(nr_aws_create_metric) {
+  if (NULL != wraprec) {
+    lib_aws_sdk_php_add_supportability_metric(wraprec->classname);
+  }
+
+  NR_PHP_WRAPPER_CALL;
+}
+NR_PHP_WRAPPER_END
+
+NR_PHP_WRAPPER(nr_aws_version) {
+  (void)wraprec;
+
+  lib_aws_sdk_php_handle_version();
+  NR_PHP_WRAPPER_CALL;
+}
+NR_PHP_WRAPPER_END
+
+/*
+ * The ideal file to begin immediate detection of the aws-sdk is:
+ * aws-sdk-php/src/functions.php
+ * Unfortunately, Php8.2+ and composer autoload leads to the
+ * file being optimized directly and not loaded.
+ *
+ * Options considered:
+ *
+ * 1. for PHP8.2, and only optimizable libraries, when encountering autoload.php
+ * files, ask the file what includes it added and check against only the
+ * optimizable library. Small overhead incurred when encountering an autoload
+ * file, but detects aws-sdk-php immediately before any sdk code executes
+ * (changes needed for this are detailed in the original PR)
+ * 2. use a file that gets called later and only when AwsClient.php file file is
+ * called. It's called later and we'll miss some instrumentation, but if we're
+ * only ever going to be interested in Client calls anyway, maybe that's ok?
+ * Doesn't detect Sdk.php (optimized out) so when customers only use that or
+ * when they use it first, we will not instrument it. This only detects when a
+ * Client is called to use a service so potentially misses out on other
+ * instrumentation and misses out when customers use the aws-sdk-php but use
+ * non-SDK way to interact with the service (possibly with redis/memcached).
+ * This way is definitely the least complex and lowest overhead and less
+ * complexity means lower risk as well.
+ * 3. Directly add the wrappers to the hash map. With potentially 50ish clients
+ * to wrap, this will add overhead to every hash map lookup. Currently
+ * implemented option is 2, use the AwsClient.php as this is our main focus.
+ * This means until a call to an Aws/AwsClient function,
+ * all calls including aws\sdk calls are ignored.  Version detection will be
+ * tied to Aws/ClientResolver::_apply_user_agent which is ALWAYS called when
+ * dealing with aws clients.  It will not be computed from
+ * Aws/Sdk::__constructor which would at best be duplicate info and worst would
+ * never be ignored until a client is called.
+ */
+void nr_aws_sdk_php_enable() {
+  /* This is will guarantee we can extract the version. */
+  nr_php_wrap_user_function(NR_PSTR("Aws\\ClientResolver::_apply_user_agent"),
+                            nr_aws_version);
+  /* This is the base of all other Clients */
+  nr_php_wrap_user_function(NR_PSTR("Aws\\AwsClient::__construct"),
+                            nr_aws_create_metric);
+  /* The Sdk class */
+  nr_php_wrap_user_function(NR_PSTR("Aws\\Sdk::__construct"),
+                            nr_aws_create_metric);
+
+  nr_php_wrap_user_function(NR_PSTR("Aws\\S3\\S3Client::__construct"),
+                            nr_aws_create_metric);
```

Comment:
Use `_extra` variant of `nr_php_wrap_user_function` to store service name in `wraprec` and use it in the metric name instead of class name:
```suggestion
nr_php_wrap_user_function_extra(NR_PSTR("Aws\\S3\\S3Client::__construct"),
                            nr_aws_create_metric, "S3");
```
This gives more control over how long the metric name can be and therefore allows for more predictable results of `lib_aws_sdk_php_add_supportability_metric` (it will less likely truncate anything). With this change, the implementation of `nr_aws_create_metric`  should be changed to:
```
NR_PHP_WRAPPER(nr_aws_create_metric) {
  if (NULL != wraprec && NULL != wraprec->extra) {
    lib_aws_sdk_php_add_supportability_metric(wraprec->extra);
  }

  NR_PHP_WRAPPER_CALL;
}
NR_PHP_WRAPPER_END
```

Reply:
Please see: [7d98b70](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/7d98b70ca2e9ada5dc79739fefa63e535ae6220d)

Instrumented function mostly did the job, but would require wrapping 50ish Aws Clients variations (and would need to determine on case by case basis what to use, for instance, dynamodbclient::__construct does not show up for PHP 8.2+).

Now we wrap the function that the aws-sdk-php uses to determine the service whenever any client is created and create a metric based off the name we snag from that. This automatically will create metrics for all clients.

metric prefix is Supportability/PHP/AWS/Services/%s/
(detected is not required. it's only needed for frameworks to differentiate with forced).